platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/Providers.java

```diff
@@ -93,6 +95,69 @@ public static <T> ProviderInternal<T> changing(SerializableCallable<T> value) {
         return new ChangingProvider<>(value);
     }
 
+    public static <T> Provider<T> memoizing(ProviderInternal<T> provider) {
+        return memoizing(provider, null);
+    }
+
+    public static <T> Provider<T> memoizing(ProviderInternal<T> provider, @Nullable SerializableSupplier<DisplayName> displayName) {
+        return new MemoizingProvider<>(provider, displayName);
+    }
+
+    public static class MemoizingProvider<T> extends AbstractMinimalProvider<T> {
+        private final ProviderInternal<T> provider;
+        @Nullable
+        private Value<? extends T> value;
+        @Nullable
+        private final Supplier<DisplayName> displayName;
+
+        public MemoizingProvider(ProviderInternal<T> provider, @Nullable Supplier<DisplayName> displayName) {
+            this.provider = provider;
+            this.displayName = displayName;
+        }
+
+        @Override
+        protected @Nullable DisplayName getDeclaredDisplayName() {
+            return displayName != null ? displayName.get() : null;
+        }
+
+        @Override
+        public ExecutionTimeValue<? extends T> calculateExecutionTimeValue() {
+            if (value != null) {
+                return ExecutionTimeValue.value(value);
+            }
+            ExecutionTimeValue<? extends T> executionTimeValue = provider.calculateExecutionTimeValue();
+            if (executionTimeValue.isMissing()) {
+                return executionTimeValue;
+            }
+            if (executionTimeValue.hasFixedValue()) {
+                value = executionTimeValue.toValue();
+                return ExecutionTimeValue.value(value);
```

Comment:
🤔 We'll discard the display name when going through CC. Not sure if that's a big issue, considering the use case: the report will have names, only carrying the obtained provider to execution and trying to display it there will highlight the problem.